6699 problems with v2 thesauri

mfacar opened the pull request:
fixes #6699
fixes #6696
fixes #7043
fixes #6959
fixes #6676


PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

mfacar committed:
using table v2

mfacar committed:
using table v2

mfacar committed:
code covered by tests

mfacar committed:
rename test

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
basic e2e tests

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri

mfacar committed:
update yarn lock

mfacar committed:
Merge branch '6699-problems-with-v2-thesauri' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
updates yarn lock

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
extended e2e thesauri tests

mfacar committed:
removes death code

mfacar committed:
ensures selector

mfacar committed:
remove obsolete translation

mfacar committed:
clean commands

mfacar committed:
removes empty groups

Zasa-san commented:
First round of functional review. This is focused solely on the thesauri screen:

- Long labels look odd with the parenthesis on the 'translated'. Super minor thing, I think we can just leave it like that until we have a global solution for multiline values on tables. But maybe it's a quick fix.

![image](https://github.com/user-attachments/assets/11bd2dad-8895-4d9f-98ac-834eb2d7bd14)

- When you are adding a group you cannot create it empty. Is this intended? Considering that now empty groups have the placeholder row to add items to it.

- You can drag all the items out of a group and still save the thesauri. This makes the group into a root item. Is this intended? Should the validation fail or the group get deleted? (If you leave the `subRows` property as an empty array it should still register it as a group on the table).

- Editing a group with values is odd, because it lists all the thesauri inside it in the sidepanel. Maybe when you edit a group just show the group label? We are not offering ways of deleting the items inside, or sorting them, so makes no sense to have them there.

- Edit a group -> Change the name of one of the items inside it, or even the label of the group itself -> Save and close the sidepanel -> Open the sidepanel again. It’s empty.

- Expan a group -> edit a child -> save and close the sidepanel -> the change is applied to the group label, not the child label.

Zasa-san commented:
Fiddling around with thessauri I see some inconsistencies in the `id` property of the database. Some of them have a long format `id`, other have what appears to be one of the client's ID generated codes. 

![image](https://github.com/user-attachments/assets/e27796aa-e5ce-422d-ac75-f4a1398f90b1)

This is a thesauri created in development. The `id` property is consistent:

![image](https://github.com/user-attachments/assets/edb8bc64-b890-43b4-a517-b38365af5e22)

mfacar committed:
edit child

mfacar committed:
fixes from review

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
clear import field

mfacar committed:
restores form reset by the moment

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/ThesaurusForm.tsx:
```diff
@@ -1,212 +1,70 @@
-  const addGroupSubmit = (group: FormThesauriValue) => {
-    const thesaurusValuesCopy = thesaurusValues.filter(item => item._id !== group._id);
-    setThesaurusValues([...thesaurusValuesCopy, group]);
+    navigate(`../edit/${savedThesaurus._id}`);
```
Is this to reload the page? We have something similar here 
https://github.com/huridocs/uwazi/blob/cde642be4baffd54d9876c182d229ff08d58a900/app/react/V2/Routes/Settings/Pages/PageEditor.tsx#L86

The example also prevents a problems where if the user navigates back they would end up in the new thesaurus screen in some cases.

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/components/TableComponents.tsx:
```diff
@@ -1,46 +1,39 @@
-  </div>
-);
+const ThesaurusValueLabel = ({ getValue, cell }: CellContext<ThesaurusValueSchema, string>) => {
+  const { thesaurus } = cell.getContext().column.columnDef.meta!.data;
```
Is this info nowhere on the `row`?

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/components/TableComponents.tsx:
```diff
@@ -75,23 +68,30 @@ interface TableThesaurusValue extends ClientThesaurusValue, Omit<ClientThesaurus
-    meta: { headerClassName: 'w-11/12' },
-  }) as ColumnDef<TableThesaurusValue, 'label'>,
+    meta: { headerClassName: 'w-11/12', data: { thesaurus } },
+  }) as ColumnDef<ThesaurusRow, 'label'>,
```
The type should be inferred automatically. No need to cast it maybe?

mfacar replied:
Totally!

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/components/ThesauriTable.tsx:
```diff
@@ -0,0 +1,77 @@
+  const navigateToEditThesaurus = (thesaurus: Row<ThesauriRow>) => {
+    navigate(`./edit/${thesaurus.original._id}`);
+  };
+  const columnHelper = createColumnHelper<any>();
```
You can use the type here, and avoid casting in every column definition below

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/specs/Thesauri.spec.tsx:
```diff
@@ -0,0 +1,469 @@
+    );
+
+    const renderComponent = () => {
+      const reduxStore = createMockStore([thunk])(() => ({
```
If we incorporate this method of testing routes, the mock store default basic values should probably be extracted into the shared folder for V2 ``V2/shared`

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/specs/Thesauri.spec.tsx:
```diff
@@ -0,0 +1,469 @@
+                },
+              ],
+            },
+            undefined
```
Is this undefined in the expect correct?

mfacar replied:
It was expected because the headers are not passed to the calls of the API from the component, however, the methods are not called anywhere else so I removed the param in the API client.

Zasa-san replied:
Ok, if we are sure those headers are never passed, we can remove them.

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/specs/Thesauri.spec.tsx:
```diff
@@ -0,0 +1,469 @@
+    });
+
+    describe('Edit Thesaurus', () => {
+      it('should edit a root item', async () => {
```
There are some tests that don't have any `expect`s. Is that correct? How do you check if the test fails?

mfacar replied:
It's a good one, I'm using screen.get... functions which fail in case the looken element doesn't exists in the document (a couple use screen.query... which doesn't fail if the element doesn't exist), also @testing-library/react/pure to keep the state between the tests, but adding expects definitively helps a lot with the readability of the tests, I added a function to check the right call of the save function where it applies.

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/specs/fixtures.ts:
```diff
@@ -0,0 +1,46 @@
+  {
+    _id: 'thesaurus1',
+    name: 'Colors',
+    values: [{ _id: 't1_black', label: 'black', id: 'id1_black' }],
```
I don't think thesaurus values have `_id`. Just mentioning in case some test relies on it being there.

mfacar replied:
I'm adding the _id to ensure the values are kept through the sanitization functions, this because #905 & #4053  suggest there could be values with _id, the component is only handling id given that that currently that is the property generated by the endpoint.

Zasa-san reviewed (changes requested):
Lengthy review process, but it's looking great! 
I have some doubt's about test using `_id` in the fixtures for a thesaurus values. If there are cases where the client is inserting those `_id` in the values we should check to avoid putting them in rather than sanitizing them. (Unless I'm mistaken thesaurus values would not haven an `_id` property ever).

mfacar committed:
code review fixes

mfacar commented:
Thank you @Zasa-san! 
> * Long labels look odd with the parenthesis on the 'translated'. Super minor thing, I think we can just leave it like that until we have a global solution for multiline values on tables. But maybe it's a quick fix.
- In the meantime, I added a border instead of the parenthesis.
> * When you are adding a group you cannot create it empty. Is this intended? Considering that now empty groups have the placeholder row to add items to it.
- In fact, an empty group could not be used as a valid option of a template property, a validation in the endpoint is also required, I created #7144 regarding to that.
> * You can drag all the items out of a group and still save the thesauri. This makes the group into a root item. Is this intended? Should the validation fail or the group get deleted? (If you leave the `subRows` property as an empty array it should still register it as a group on the table).
- I disabled the group edition due to #6906 and #7144.  By the moment if all the items of a group are deleted by selection, the group is also removed.
> * Editing a group with values is odd, because it lists all the thesauri inside it in the sidepanel. Maybe when you edit a group just show the group label? We are not offering ways of deleting the items inside, or sorting them, so makes no sense to have them there.
- I'm not sure about this, although it is different from other places where items can be deleted, you can also make a massive update of all the members without clicking each row. @juanmnl any thoughts? 
> * Edit a group -> Change the name of one of the items inside it, or even the label of the group itself -> Save and close the sidepanel -> Open the sidepanel again. It’s empty.
- Fixed
> * Expan a group -> edit a child -> save and close the sidepanel -> the change is applied to the group label, not the child label.
- OMG!, Thanks for this caught, it's fixed.

mfacar commented:
> Fiddling around with thessauri I see some inconsistencies in the `id` property of the database. Some of them have a long format `id`, other have what appears to be one of the client's ID-generated codes.

For the record, @Zasa-san found the different formats for id are caused by adding values into a thesaurus from the metadata editor.   There is some legacy code in app/react/Thesauri/actions/thesauriActions.js that should be removed when we rewrite the related components.

mfacar committed:
extracts function to reduce complexity

mfacar committed:
update of snapshots due parenthesis

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri

mfacar committed:
conditional translate

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
Merge branch '6699-problems-with-v2-thesauri' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
relies on endpoint for id generation

mfacar committed:
removes unused import

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/components/ThesauriTable.tsx:
```diff
@@ -0,0 +1,77 @@
+  const navigateToEditThesaurus = (thesaurus: Row<ThesauriRow>) => {
+    navigate(`./edit/${thesaurus.original._id}`);
+  };
+  const columnHelper = createColumnHelper<ThesauriRow>();
```
I believe that the columnHelper and the columns themselves can be moved outside the component, they don't use anything from the component (only `edit` and that's passed when the columns are built in :59). I'm not sure how much it impacts component render time to rebuild all of this on every rerender, but it's safe to declare it outside. 

You could even momize the call to the column builder function and have it depend on the edit parameter. But maybe that's no so necessary.

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/specs/helpers.spec.ts:
```diff
@@ -0,0 +1,464 @@
+  thesaurusAsRow,
+} from '../helpers';
+
+const thesaurusValues: {}[] = [
```
Just an idea, feel free to ignore it: use the correct type here and if you have type issues ts-ignore them with a message that explains why those incorrect values are there. Like _id is there to check correct sanitization, or why is the empty value array. Could be helpful for the next dev that has to come here.

Also useful for sanitizedThesaurusValues so that we are sure it's the correct expected values.

Zasa-san commented on app/react/V2/Routes/Settings/Thesauri/EditThesaurus.tsx:
```diff
@@ -0,0 +1,220 @@
+                <Button styling="light" onClick={addGroup}>
+                  <Translate>Add group</Translate>
+                </Button>
+                <Button styling="light" onClick={sortValues(thesaurusValues, setThesaurusValues)}>
```
The table has it's own `sortingFn` prop that can be used to pass a custom sorting function. Could this not be used instead of a custom sorting function? Just rely on the table to run the sorting by title when this button is clicked? `app/react/V2/Routes/Settings/IX/IXSuggestions.tsx:252` does something like it.

mfacar replied:
@Zasa-san I agree with you, but my attempt to use the sortingFn to update the sortingState from the table client turned into a more complex option, for sure I'm missing some part of the configuration, so for now I'll keeping the external sort function

Zasa-san commented on cypress/e2e/settings/thesauri.cy.ts:
```diff
@@ -20,104 +25,212 @@ describe('Thesauri configuration', () => {
+    cy.get('.only-desktop a[aria-label="Settings"]').click();
     cy.contains('span', 'Thesauri').click();
-    cy.get('tbody tr:nth-of-type(1) input').should('be.disabled');
+    cy.contains('New Thesaurus')
```
I think you can do a `cy.contains('tr', 'New Thesaurus').contains('button', 'Edit').click();` here

mfacar replied:
Certainly, much simpler!

Zasa-san commented on cypress/e2e/settings/thesauri.cy.ts:
```diff
@@ -1,11 +1,16 @@
 /* eslint-disable max-statements */
```
The e2e snapshots fail locally for me. If they are green on the CI, I'm ok. It's a bit difficult for me to check that the e2e covers all the expected flows correctly, although.

mfacar replied:
I have the same problem locally, the problem is reported in Cypress and happens when using Chrome as the browser, at the moment running with Electron or the console is the alternative to check the tests.

Zasa-san reviewed (approved):
Amazing!

The remaining comments are suggestions, but generally I'm ok with merging as is.

mfacar committed:
simplified e2e get of actions

mfacar committed:
Merge branch 'development' of https://github.com/huridocs/uwazi into 6699-problems-with-v2-thesauri

mfacar committed:
refactor from code review

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri

mfacar committed:
fix broken component

mfacar committed:
use uuid for value id

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri

txau commented:
The problem with importing the CSV still persists. 

![image](https://github.com/user-attachments/assets/9cd4cec7-a59e-4f63-8f6c-9a78cc7a3c5c)

txau reviewed (approved):
LGTM

mfacar committed:
Merge branch 'development' into 6699-problems-with-v2-thesauri
